Runtime/Triangulator.cs

```diff
@@ -283,49 +288,485 @@ public void Dispose()
+            public LengthSq CircumRadiusSq(Coord a, Coord b, Coord c);
+            public Coord CircumCenter(Coord a, Coord b, Coord c);
+            public int Orient2dFast(Coord a, Coord b, Coord c);
+            public bool InCircle(Coord a, Coord b, Coord c, Coord p);
+            public bool PointInsideTriangle(Coord p, Coord a, Coord b, Coord c);
+            public bool PointLineSegmentIntersection(Coord a, Coord b0, Coord b1);
+            public bool EdgeEdgeIntersection(Coord a0, Coord a1, Coord b0, Coord b1);
+            public bool IsConvexQuadrilateral(Coord a, Coord b, Coord c, Coord d);
+
+            public bool IsValidCoordinate(Coord coord);
+            public bool Equals (Coord a, Coord b);
+            public float Distance (Coord a, Coord b);
+            public LengthSq DistanceSq(Coord a, Coord b);
+            public bool LengthSmaller (LengthSq lhs, LengthSq rhs);
+            public LengthSq MultiplyLengthSq (LengthSq lhs, float multiplier);
+
+            public AffineTransform2D CalculatePCATransformation(NativeArray<Coord> positions);
+            public AffineTransform2D CalculateLocalTransformation(NativeArray<Coord> positions);
+
+            public void Transform (AffineTransform2D transform, NativeArray<Coord> points);
+            public void Transform (AffineTransform2D transform, [NoAlias] NativeArray<Coord> points, [NoAlias] NativeArray<Coord> outPoints);
+            public void InverseTransform (AffineTransform2D transform, [NoAlias] NativeArray<Coord> points, [NoAlias] NativeArray<Coord> outPoints);
+            public LengthSq InfLengthSq { get; }
+
+            public int HashKey(Coord p, Coord origin, int hashSize);
+            public void BoundingBox (ReadOnlySpan<Coord> positions, out Coord min, out Coord max, out Coord center);
+            public Coord Lerp(Coord a, Coord b, float t);
+            public bool SmallestInnerAngleIsBelowThreshold(Coord a, Coord b, Coord c, float angleThreshold);
+        }
+
         #region Jobs
+
+        public struct InputData<Coord> where Coord : unmanaged
+        {
+            public NativeArray<Coord> Positions;
+            [NativeDisableContainerSafetyRestriction]
+            public NativeArray<int> ConstraintEdges;
+            [NativeDisableContainerSafetyRestriction]
+            public NativeArray<Coord> HoleSeeds;
+        }
+
+        public struct OutputData<Coord> where Coord : unmanaged {
+            [NativeDisableContainerSafetyRestriction]
+            public NativeList<Coord> Positions;
+            public NativeList<int> Triangles;
+            public NativeReference<Status> Status;
+        }
         [BurstCompile]
-        private struct TriangulationJob : IJob
+        private struct TriangulationJob<Coord, LengthSq, CoordinateUtils> : IJob where Coord : unmanaged where LengthSq : unmanaged, IComparable<LengthSq> where CoordinateUtils : ICoordinateUtils<Coord, LengthSq>, new()
```

Comment:
I would rename generic parameters name with the following

| Current  | My proposal |
| ------------- | ------------- |
| `Coord ` | `TCoords` |
| `LengthSq`  | `TLengthSq` |
| `CoordinateUtils` | `TUtils` |

Reply:
Hi

The primary reason I really do not like a class based api is that it makes it impossible to use it from within burst compiled code. It also creates unnecessary GC allocations, but that's only a minor issue.

I also don't really get any aesthetic or functional reason for choosing it?

```
using var t2 = new Triangulator(Allocator.Persistent) { Settings = { RefineMesh = true }, Input = { Positions = positions } };
t2.Run();
var output = t2.Output;
```

just seems worse, or at least equivalent to:

```
var output = new Triangulator.Output(Allocator.Persistent);
Triangulator.Triangulate({ RefineMesh = true },  { Positions = positions }, output).Run();
```


(we can make the Triangulate method return the whole job struct, to allow the user to call Schedule or Run as on the object, as desired, though this comes with a minor scheduling cost due to having to copy the large struct one more time)

Reply:
I commented regarding reasons for this [here](https://github.com/andywiecko/BurstTriangulator/pull/117#discussion_r1598486191).

About GC alloc this is true only when one call `new()` every frame, but I assume that user should be aware that managed objects have cost.